src/pkcs15init/pkcs15-lib.c

```diff
@@ -1674,6 +1675,8 @@ sc_pkcs15init_generate_key(struct sc_pkcs15_card *p15card, struct sc_profile *pr
 		/* can be allocated in driver-specific generate_key() */
 		free(pubkey_args.key.u.ec.ecpointQ.value);
 		pubkey_args.key.u.ec.ecpointQ.value = NULL;
+		if (r < 0 && alg_id_free)
+			free(pubkey->alg_id);
```

Comment:
1) I think this should not be in the condition `(algorithm == SC_ALGORITHM_EC)` as this can happen also for other algorithms.
2) I think in regards to the recent changes to this code I did with valgrind, I think the `alg_id` should be copied in all cases and free here in all cases to simplify the memory management, similarly as done with the other parts of pubkey structure.